### ISSUE TITLE
Feature: Graph - Weight Only Increasing Toggle

### ISSUE THREAD

**Issue by mehalter** [graphefruit/Beanconqueror#308]
I am wondering if it would be a good idea to add a toggle to make the scale weight graph only increase over time. When I do something like a pour over I like to do a swirl with the bed of coffee and to decrease the noise communicated through the scale I have been just lifting up the entire brewer, doing the swirl and setting it down. This results in something like the picture below, as you can see if makes the weight go down but in a way that I don't necessarily care about when viewing the final graph and makes the part I am interested in much smaller.

I'm thinking this could be implemented as a smart scale setting in the settings menu and the feature itself seems like this could be done in a few different ways. Either in the data collection we don't store the decreasing values or what may be better is just in the rendering of the graph so we don't lose information that may be useful (?) in the future.

![pourover flow](https://user-images.githubusercontent.com/1591837/150539938-fb8cea93-2481-475f-8d82-551395179bda.jpg)

**Comment by graphefruit** [comment]
HI @mehalter,

adding this feature is tricky.

If you want to remove the up&down scales, the weight graph would have a delay of 1 second, because you need to remove the anomalies (which is done in the flow graph actually)
Because you need to check if there is a hiccup with the weight.

Secondly:
I'd like to work with a feature flag, and still store this data.

If the weight is not checked, you can read out of the flow graphic e.g. when was first pour, when second, how long did you swirl (because of ups and downs on the weight level).
This would be also includeable in the flow graphic share #300.

So to grab it all together:
If up&down shall be just "up", there will be a delay of weight with 1 second behind

**Comment by mehalter** [comment]
Makes sense and definitely don't want any sort of delay. Also for something like this it could just be a feature just in the viewing a brew that re-renders the graph so we wouldn't have any delays during the collection. Also that is a good point where you can tell how long the swirl happened and when these things occurred.

**Comment by graphefruit** [comment]
Yes, I mean it would be possible to make it non-laggy (you need to render the values, and then flat out afterwards and update the old indexes) and thats to be honest to much work for me right now without good outcome.

Btw: In the latest code on the actual upgrade branch, you can export the flow graphic, with white background, and also you can zoom into the graphic and pan the x axis

**Comment by mehalter** [comment]
Yeah I was checking that out this morning. Great work! I tested them out and both features working perfectly on my device

**Comment by graphefruit** [comment]
I just checked the flow graph... it isn't updated in realtime, it already has a 1 sec delay (lol)....

**Comment by graphefruit** [comment]
Nobody found out yet :>
I just trigger the `update` after all things has been checked, so the weight graph is filled with "realtime" data, but shown 1 sec later

**Comment by mehalter** [comment]
Hm, @graphefruit I just brewed with the new changes to the zooming and background stuff and I think something is broken. the flow chart doesn't refresh when I'm brewing anymore but the scale is definitely communicating data, i.e. the weight and flow numbers above the graph update and so does the water calculations. Once the brew is done and I save and open it up for viewing the correct graph is displayed, I just get no live updating graph during the brew anymore.

**Comment by mehalter** [comment]
Also note, this is before the merging of the new scale (this commit: [788b91d68359df5b80632d9cd9cf8b9605aa5590](https://github.com/graphefruit/Beanconqueror/tree/788b91d68359df5b80632d9cd9cf8b9605aa5590))

We mentioned those features in this issue so I dropped this information here, let me know if you want me to move it into a new issue for easier tracking.

**Comment by graphefruit** [comment]
While brewing you shouldn't zoom.
You would need to reset via button or something like this that the graph is working normal

**Comment by graphefruit** [comment]
Maybe this feature should just be available on detail page?

**Comment by mehalter** [comment]
I wasn't trying to zoom, the flow graph just isn't working at all when I brew

**Comment by graphefruit** [comment]
Oh sorry, so i missunderstood.

I gonna be back in about 1 1/2 hour and check the actual implemention code.
I played to much with the graphs actually :<

**Comment by mehalter** [comment]
No problem, I didn't notice earlier because I was just viewing the brews and not actively brewing until just recently. Thanks for taking a look :)

**Comment by graphefruit** [comment]
> Yeah I was checking that out this morning. Great work! I tested them out and both features working perfectly on my device

I checked in changes, and compiled.
The flowgraph worked this day again

**Comment by mehalter** [comment]
Thank you! :D

**Comment by Rescla** [comment]
Looks great! One suggestion I'd like to make is to (add an option to) hide weight values below 0g.

When looking at the graph as it is it is kind of hard to read the scale for weight. If the brewer is taken off the scale during the brew it will cause a large negative value to stretch the x scale, causing the readability of the graph to go down. I think there isn't much of a difference between showing those values as 0, or showing them as -600, assuming the scale is zeroed at the start.

The upside of this over filtering only upwards values is that it does not cause any delay. And it works if someone ever decides removing water during the brew is a good idea. The downside is that it does not solve the jitter issue, but I think that wouldn't be solved by ignoring decreasing values either. You'd probably need some kind of smoothing or detect swirling somehow.

**Comment by graphefruit** [comment]
I've had some time to think about this yesterday.
Ignoring or hiding values below 0 would just solve some of the issues.

If you take of the server and lets say you have 600g on the scale actually.
The weight would drop like:

Second 0.1: 550
Second 0.2: 420
Second 0.3: 200
Second 0.4: -400

So you'd see a dropping line, which would stop then on "0", but it would still fall down on the chart for some milliseconds.
Therefore it also cant be done realtime.
It can be done with 1 sec delay - like the actual graph is already working and noone wondered ;)

**Comment by Rescla** [comment]
I personally wouldn't mind seeing bit a dropping line. My main problem was with the X axis scaling, not so much with showing the values themselves.

If the 1 second delay is an issue, and you still want to see a strict drop off,  you could also change the measurements retroactively if you determine a 'take off event'.

I'm not sure how much of an issue a 1 second delay would be though. The scale measurement itself has some delay already, so it might be that 1s delay on top of that is just fine. I can see it being one of those things that people complain about if they know though :P

**Comment by graphefruit** [comment]
I've added two new settings values for testing on first place.
First one: Ignore negative values 
Second one: Ignore anomaly values

I've added the logic to calculate the whole second.
Check 1:
If the check detects a negative value in all weights in one second -> Negative value

Check 2:
If the check detects up and down moving values, or negative value -> Anomaly.

The value will be dropped to "null".
I search for the latest right added value and replace "null" with this one.

Funny one :>

**Comment by mehalter** [comment]
Finally got around to testing this out a little. So far I've only tested the anomaly option and it's working great! 
![brew](https://user-images.githubusercontent.com/1591837/152030918-1f8bf66d-0f2e-494a-a715-dfabdebe2da0.jpg)

This was with a similar brew to the first image I sent with 2 times where I picked it up and gave the brewer a swirl.

**Comment by graphefruit** [comment]
I'm actually playing with another chartjs plugin, maybe I need to dig a bit more here

**Comment by graphefruit** [comment]
I've made the first realtime graph sample just in #319  @mehalter  - If you activate the anomaly and the negative value the graph will update every 1 sec, if you deactivate those, the graph is realtime, and the flow is readded every second

**Comment by mehalter** [comment]
:O that's awesome! Will check it out and test it over the next few days :)

**Comment by graphefruit** [comment]
Version 6.1 is nearly finished.
I'm going to close this issue after it has been build in.
If there are issues, please reopen a new ticket
